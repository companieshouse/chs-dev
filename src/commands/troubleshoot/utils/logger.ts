import { simpleColouriser } from "../../../helpers/colouriser.js";

export default class TroubleshootOutputLogger {

    private readonly logger: (msg: string) => void;

    constructor (logger: (msg: string) => void) {
        this.logger = logger;
    }

    logMessage (message: string, stat: "Success" | "Error"): void {
        stat === "Success"
            ? this.logger(`${simpleColouriser("SUCCESS!", "bold-green")} ${message}`)
            : this.logger(`${simpleColouriser("FAILED!", "red")} ${message}`);

    }
}
