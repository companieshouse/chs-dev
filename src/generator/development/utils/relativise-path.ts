import { dirname, join, relative } from "path";

/**
 * Creates relative paths from the destination directory to the source for the
 * supplied paths supplied.
 * @param paths Array of strings or string representing the paths being
 *      converted to relative paths
 * @param source original location from which the path was relative to prior
 * @param destinationDirectory Location from which the new path will be
 *      relative from
 * @returns relative path from the destination directory to the paths provided
 */
export const relativiseServicePath = (
    paths: string | string[],
    source: string,
    destinationDirectory: string
) => {
    if (Array.isArray(paths)) {
        return paths.map(
            singleValue => relativiseServicePath(singleValue, source, destinationDirectory)
        );
    }

    const relativePathToSource = relative(destinationDirectory, source);

    return join(dirname(relativePathToSource), paths);
};

export default relativiseServicePath;
