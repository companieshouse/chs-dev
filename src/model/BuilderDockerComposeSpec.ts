/**
 * Represents a Docker Compose specification for a builder
 */
export interface BuilderDockerComposeSpec {
    /**
     * Name of the builder
     */
    name: string;

    /**
     * Version of the builder
     */
    version: string;

    /**
     * String representation of the docker compose spec - may contain variables
     * in '< >' which are removed when the docker compose file is created
     */
    builderSpec: string;
}

export default BuilderDockerComposeSpec;
