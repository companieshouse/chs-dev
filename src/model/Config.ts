/**
 * Represents project configuration defining project related properties for
 * customising the behaviours of chs-dev
 */
export type Config = {

    /**
     * name to refer to the project/environment
     */
    readonly projectName: string;

    /**
     * path to the project repository
     */
    readonly projectPath: string;

    /**
     * path to the chs-dev local repository
     */
    readonly chsDevPath?: string;

    /**
     * path to the chs-dev data
     */
    readonly chsDevDataDir?: string;

    /**
     * additional environment variables for running the environment. Values can
     * contain 'file://' prefixes will try to read file in and set as value in
     * its place.
     */
    readonly env: Record<string, string>;

    /**
     * when supplied defines the threshold for number of hours after which to
     * attempt ecr login
     */
    readonly performEcrLoginHoursThreshold?: number;

    /**
     * specifies the supported version of the CLI by the project
     */
    readonly versionSpecification?: string;

}

export default Config;
