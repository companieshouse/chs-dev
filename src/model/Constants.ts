/**
 * Constant values for chs-dev projects
 */
export const CONSTANTS = {
    /**
     * The initial Docker Compose Spec which forms the basis of the generated
     * Docker Compose specification defining the environment
     */
    BASE_DOCKER_COMPOSE_FILE: "services/infrastructure/docker-compose.yaml",

    /**
     * Value for a boolean label to be considered true
     */
    BOOLEAN_LABEL_TRUE_VALUE: "true",

    /**
     * Default value for the number of hours between ecr login checks
     */
    DEFAULT_PERFORM_ECR_LOGIN_HOURS_THRESHOLD: 8,

    /**
     * The prefix for Companies House Java Package names
     */
    JAVA_BASE_PACKAGE: "uk.gov.companieshouse",

    /**
     * The default Spring Boot dependencies for a new service
     */
    SPRING_BOOT_DEPENDENCIES: [
        "web",
        "data-mongodb"
    ],

    /**
     * The URL for the Spring Initializr service
     */
    SPRING_BOOT_INITIALIZR_URL: "https://start.spring.io",

    /**
     * The Java version for a new service
     */
    JAVA_VERSION: "21",

    /**
     * The type of Spring Boot Java project for a new service
     */
    JAVA_PROJECT_TYPE: "maven-project",

    /**
     * The version of Spring Boot for a new service
     */
    SPRING_BOOT_VERSION: "3.3.5",

    /**
     * The name of the GitHub organisation for Companies House
     */
    COMPANIES_HOUSE_GITHUB_ORGANISATION_NAME: "companieshouse",

    /**
     * The name of the repository for new service specs
     */
    NEW_SERVICE_SPECS_REPOSITORY_NAME: "proof-of-concept",

    /**
     * The branch of the repository for new service specs
     */
    NEW_SERVICE_SPECS_REPOSITORY_BRANCH: "feature/new-service-requests-concourse-based",

    /**
     * The path to the new service specs in the repository
     */
    NEW_SERVICE_SPECS_PATH: "repositories/",

    /**
     * Names of internal scrum teams
     */
    SCRUM_TEAMS: [
        "Aardvark",
        "Apollo",
        "Coblyn",
        "Element",
        "Parental Advisory",
        "Phoenix",
        "Spartacus",
        "Thundercats",
        "Titans",
        "Toro Loco"
    ],

    SERVICE_NAMES: [
        "Filing",
        "Incorporate",
        "Close"
    ]
};

export default CONSTANTS;
