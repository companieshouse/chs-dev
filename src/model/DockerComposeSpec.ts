type Secret = {
    environment: string
} | {
    file: string
}

/**
 * Represents the Docker Compose Spec file when read in from yaml
 */
export type DockerComposeSpec = {
    services: {
        [serviceName: string] : {
            labels?: string[] | Record<string, string>,
            environment?: string[] | Record<string, string>,
            image?: string,
            build?: Record<string, any>,
            env_file?: string | string[],
            depends_on?: string[] | Record<string, Record<string, any>>
            secrets?: string[] | Record<string, any>,
            ports?: string | string[]
        } & Record<string, any>
    },
    secrets?: Record<string, Secret>,
    include?: string[],
    networks?: Record<string, any>,
    volumes?: Record<string, any>
};
