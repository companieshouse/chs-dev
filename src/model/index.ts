export enum LogCoverage {
  WATCH,
  LOG,
}

export enum Prune {
  VOLUME,
  NETWORK,
  IMAGE,
  CONTAINER,
  BUILDER,
}
