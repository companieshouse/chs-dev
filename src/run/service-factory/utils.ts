import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
