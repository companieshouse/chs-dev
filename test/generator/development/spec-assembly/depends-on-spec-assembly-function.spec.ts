import { expect } from "@jest/globals";
import dependsOnSpecAssemblyFunction from "../../../../src/generator/development/spec-assembly/depends-on-spec-assembly-function";
import { DockerComposeSpec } from "../../../../src/model/DockerComposeSpec";
import { services } from "../../../utils/data";
import { generateServiceSpec } from "../../../utils/docker-compose-spec";

describe("dependsOnSpecAssemblyFunction", () => {

    it("adds dependencies to development spec from service spec", () => {
        const service = services[6];
        const serviceSpec = generateServiceSpec(service);
        serviceSpec.services[service.name].depends_on = {
            "service-one": {
                condition: "service_healthy"
            },
            "service-two": {
                condition: "service_healthy"
            }
        };

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {
                    depends_on: {
                        [`${service.name}-builder`]: {
                            condition: "service_completed_succcessfully",
                            restart: true
                        }
                    }
                }
            }
        };

        dependsOnSpecAssemblyFunction(
            developmentDockerComposeSpec,
            {
                service,
                projectPath: "/home/test/project",
                serviceDockerComposeSpec: serviceSpec,
                builderDockerComposeSpec: {
                    name: "java",
                    version: "v1",
                    builderSpec: ""
                }
            }
        );

        expect(Array.isArray(developmentDockerComposeSpec.services[service.name].depends_on)).toBe(false);
        expect(developmentDockerComposeSpec.services[service.name].depends_on || {}).toMatchSnapshot();
    });

    it("converts from array", () => {
        const service = services[6];
        const serviceSpec = generateServiceSpec(service);
        serviceSpec.services[service.name].depends_on = [
            "service-three",
            "service-five"
        ];

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {
                    depends_on: {
                        [`${service.name}-builder`]: {
                            condition: "service_completed_succcessfully",
                            restart: true
                        }
                    }
                }
            }
        };

        dependsOnSpecAssemblyFunction(
            developmentDockerComposeSpec,
            {
                service,
                projectPath: "/home/test/project",
                serviceDockerComposeSpec: serviceSpec,
                builderDockerComposeSpec: {
                    name: "java",
                    version: "v1",
                    builderSpec: ""
                }
            }
        );

        expect(Array.isArray(developmentDockerComposeSpec.services[service.name].depends_on)).toBe(false);
        expect(developmentDockerComposeSpec.services[service.name].depends_on || {}).toMatchSnapshot();
    });

    it("does not modify depends_on when not present in service spec", () => {
        const service = services[0];
        const serviceSpec = generateServiceSpec(service);

        serviceSpec.services[service.name].depends_on = undefined;

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {}
            }
        };

        dependsOnSpecAssemblyFunction(
            developmentDockerComposeSpec,
            {
                service,
                projectPath: "/home/test/project",
                serviceDockerComposeSpec: serviceSpec,
                builderDockerComposeSpec: {
                    name: "java",
                    version: "v1",
                    builderSpec: ""
                }
            }
        );

        expect(Object.keys(developmentDockerComposeSpec.services[service.name])).not.toContain("depends_on");
    });

});
