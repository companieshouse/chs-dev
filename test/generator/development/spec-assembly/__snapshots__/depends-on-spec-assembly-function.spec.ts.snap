// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependsOnSpecAssemblyFunction adds dependencies to development spec from service spec 1`] = `
{
  "service-one": {
    "condition": "service_healthy",
  },
  "service-seven-builder": {
    "condition": "service_completed_succcessfully",
    "restart": true,
  },
  "service-two": {
    "condition": "service_healthy",
  },
}
`;

exports[`dependsOnSpecAssemblyFunction converts from array 1`] = `
{
  "service-five": {
    "condition": "service_started",
  },
  "service-seven-builder": {
    "condition": "service_completed_succcessfully",
    "restart": true,
  },
  "service-three": {
    "condition": "service_started",
  },
}
`;
